src/vset.h

```diff
@@ -0,0 +1,312 @@
+ *      - Original bucket retains its timestamp (but holds fewer entries).
+ *      - New bucket is inserted before the current one with its own tighter timestamp.
+ *
+ *        Example:
+ *
+ *        Before:
+ *             [ Entry0 ... Entry126 ]  -> B(ts=128)
+ *
+ *        After Split:
+ *             [ Entry0...Entry62 ]     -> New B(ts=64)
+ *             [ Entry63...Entry126 ]   -> Original B(ts=128)
+ *
+ * 3. **Convert to hashtable:**
+ *      - When no clean split is found (e.g. all entries share similar expiry),
+ *        and realignment is not possible.
+ *      - This allows efficient O(1) lookups even with clustered expiry values.
+ *
+ *        Vector B(ts=128) -> Hashtable B(ts=128)
+ *
+ * This hierarchical design ensures:
+ *   - Efficient memory usage (tight buckets)
+ *   - Predictable iteration by expiry time
+ *   - Low overhead insertions & deletions
+ *   - Graceful promotion & demotion of bucket types
+ *
+ * NOTE: Buckets are always sorted by their `bucket_ts` in the radix tree (RAX),
+ *       which allows efficient search for insertion/removal based on expiry.
+ *
+ *-----------------------------------------------------------------------------
+ * RAX Bucket Layout
+ *-----------------------------------------------------------------------------
+ *
+ * * RAX View with Time Keys:
+ *
+ *     expiry_buckets = rax * | 0x6
+ *
+ *     +--------------------------+
+ *     | RAX (key = bucket_ts)   |
+ *     |--------------------------|
+ *     | "000016" -> [entry1]     |  <- Vector (SINGLE->VECTOR->HT)
+ *     | "000032" -> [entry2...]  |  <- Full vector, might split
+ *     | "000048" -> [entry...]   |
+ *     +--------------------------+
+ *
+ * * Splitting a Full Vector in RAX:
+ *
+ *     Suppose vector at key "000032" has 13 entries:
+ *
+ *     1. Use binary search to find a transition point in expiry bucket_ts.
+ *        We search the first 2 following entries which belong to different lwo granularity time windows,
+ *        but as close as possible to the middle of the vector:
+ *            [entry1, entry7, ..., entry13]
+ *                          ↑
+ *                         split (first where get_bucket_ts(entry) > min_ts)
+ *
+ *     2. Create two vectors:
+ *            bucket A -> [entry1..entry6]  with key = "000032"
+ *            bucket B -> [entry7..entry13] with key = "000048"
+ *
+ *     3. Insert both back to the RAX.
+ *
+ *-----------------------------------------------------------------------------
+ * Bucket Lifecycle
+ *-----------------------------------------------------------------------------
+ *
+ *     NONE
+ *       |
+ *       v
+ *     SINGLE (1 entry)
+ *       |
+ *       v
+ *     VECTOR (sorted, up to 127)
+ *       |
+ *       v
+ *     RAX
+ *       |
+ *       v
+ *     +-------------+
+ *     | key -> bucket|
+ *     +-------------+
+ *     | "000016" -> VECTOR
+ *     | "000032" -> HT
+ *     | "000048" -> SINGLE
+ *     +-------------+
```

Comment:
The table in the bottom doesn't really show the promotions for buckets within a rax. Should it? Maybe use arrows similar to the top-level bucket promotions?

```suggestion
 *-----------------------------------------------------------------------------
 * Bucket Lifecycle
 *-----------------------------------------------------------------------------
 *
 *     NONE
 *       |
 *       v
 *     SINGLE (1 entry)
 *       |
 *       v
 *     VECTOR (sorted, up to 127)
 *       |
 *       v
 *     RAX (holds multiple buckets, keyed by each bucket's end timestamp)
 *     Bucket types within a RAX:
 *
 *                    SINGLE
 *                      |
 *                      v  
 *                    VECTOR (sorted, up to 127, can split
 *                      |     into multiple vectors)
 *                      |
 *                      v
 *                   HASHTABLE (only when a vector can't split)
 *       
```